File: python/dp/maximum_subarray.py

# 最大子序和
from typing import List


class Solution:
    def maxSubArray_1(self, nums: List[int]) -> int:
        for i in range(1, len(nums)):
            nums[i] = max(nums[i], nums[i] + nums[i - 1])
        return max(nums)

    def maxSubArray_2(self, nums: List[int]) -> int:
        res, maxim = nums[0], nums[0]
        for i in range(1, len(nums)):
            maxim = max(maxim + nums[i], nums[i])
            res = max(maxim, res)
        return res

File: go/dp/maximum_subarray.go
package main

import "math"

//最大子序和

func maxSubArray(nums []int) int {
	res, max := nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		max = int(math.Max(float64(nums[i]), float64(max+nums[i])))
		res = int(math.Max(float64(res), float64(max)))
	}
	return res
}
